src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -181,6 +199,74 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
             return MsQuicStatusCodes.Success;
         }
 
+        private static uint HandleEventPeerCertificateReceived(State state, ref ConnectionEvent connectionEvent)
+        {
+            SslPolicyErrors sslPolicyErrors  = SslPolicyErrors.None;
+            X509Chain? chain = null;
+            X509Certificate2? certificate = null;
+
+            if (!OperatingSystem.IsWindows())
+            {
+                // TBD fix validation with OpenSSL
+                return MsQuicStatusCodes.Success;
+            }
+
+            MsQuicConnection? connection = state.Connection;
+            if (connection == null)
+            {
+                return MsQuicStatusCodes.InvalidState;
+            }
+
+            if (connectionEvent.Data.PeerCertificate.PlatformCertificateHandle != IntPtr.Zero)
+            {
+                certificate = new X509Certificate2(connectionEvent.Data.PeerCertificate.PlatformCertificateHandle);
+            }
+
+            try
+            {
+                if (certificate == null)
```

Comment:
I don't expect this to ever happen. We shouldn't be calling you with "Peer Certificate Received" if we didn't have a certificate. Personally, I'd say you could just assert here or something.

Reply:
The null makes perfect sense for server side where client's certificate is optional. I just need to complete that part but the validation callback will be shared.